source/slang/parser.cpp

```diff
@@ -1004,29 +1099,62 @@ namespace Slang
         }
     }
 
-    static void ParseFuncDeclHeader(
+    static RefPtr<Decl> ParseFuncDeclHeader(
         Parser*                     parser,
         DeclaratorInfo const&       declaratorInfo,
-        RefPtr<FuncDecl>  decl)
+        RefPtr<FuncDecl>  decl,
+        RefPtr<GenericDecl> genDecl)
     {
-        parser->PushScope(decl.Ptr());
+        RefPtr<Decl> retDecl = decl;
 
         parser->FillPosition(decl.Ptr());
         decl->loc = declaratorInfo.nameAndLoc.loc;
 
         decl->nameAndLoc = declaratorInfo.nameAndLoc;
+
+        // if return type is a DeclRef type, we need to update its scope to use this function decl's scope
```

Comment:
This doesn't seem like it would be enough. The return type might be a more complex type expression that still involves the type parameters (e.g., `SomeGeneric<T>` or `T[]`, where the actual `DeclRefType` that refers to `T` isn't the top-most one).

It seems like the short-term fix here would be to recursively walk the result type expression and replace any nested scopes.

The long-term fix is probably to do away with attaching lookup scopes to terms during parsing, and instead try to do it during semantic checking. That would be a big change with a lot of chances for errors, though.

A third option (which might be hard for you to accept right now...) is to insist that generic functions should be written with an alternate function syntax entirely, so that instead of:

    T foo<T>();

we write:

    func foo<T>() -> T;